src/server/journal/journal_slice.cc

```diff
@@ -192,8 +187,11 @@ void JournalSlice::AddLogRecord(const Entry& entry, bool await) {
     DVLOG(2) << "AddLogRecord: run callbacks for " << entry.ToString()
              << " num callbacks: " << change_cb_arr_.size();
 
-    for (const auto& k_v : change_cb_arr_) {
-      k_v.second(*item, await);
+    const size_t size = change_cb_arr_.size();
```

Comment:
if we remove one and add one the size will be the same

Reply:
we should use iterator to the last element (not end)

Reply:
This won't happen. We acquire a `reader` mutex above and `UnregisterOnChange` acquires a `writer`.

```
  206 void JournalSlice::UnregisterOnChange(uint32_t id) {                                               
  207   // we need to wait until callback is finished before remove it                                   
  208   lock_guard lk(cb_mu_);   
```

Reply:
Could you clarify where we acquire  a reader mutex

Reply:
https://github.com/dragonflydb/dragonfly/pull/3381/files#diff-e743e1fb3b0c3695aacf5782428d0ebdbaec34585abab7100afc5e367d934c2bR186 

`shared_mutex` allows multiple `readers`. `if a writer, unique_lock` arrives then it has to block for all readers to finish

Reply:
3 lines above